Azure-Samples/azure-ml-federated-learning
# Pneumonia xray example

**This PR adds an example cross-silo FL pipeline for doing pneumonia detection on native Azure ML.** It is directly adapted from [that example using NVFlare](https://github.com/Azure/medical-imaging/tree/main/federated-learning). We use Azure ML computes clusters and Azure storages in various regions to mimic the case of geographically distributed data.

The entry point for running this example is the [pneumonia.md](https://github.com/Azure-Samples/azure-ml-federated-learning/blob/thomas/pneumonia-xray-example/docs/real-world-examples/pneumonia.md) file. It contains all the instructions. Here is a quick summary to help with the code review.

1. Provisioning - For simplicity, we recommend creating an open setup, just like in the quickstart.
2. Data - Get the dataset from Kaggle and split it into 3 silos by cloning the repo, adding some GitHub secrets and running a GitHub action.
3. Job - Modify the config file to point at your workspace, then run a single command.

Successful run [here](https://ml.azure.com/runs/tender_cord_v9lkcc23ry?wsid=/subscriptions/48bbc269-ce89-4f6f-9a12-c6f91fcb772d/resourcegroups/thopo-owner-fldev-rg/workspaces/aml-tp-pneumo&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#).

Closes #151 

## Purpose

* The goal with this addition is to increase the advertisement value of our repo. We are targeting the healthcare industry.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?


```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[x] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
* Get the code by **forking** the repo (not _cloning_, see more details in the instructions).
* Test the code by following the instructions in the [pneumonia.md](https://github.com/Azure-Samples/azure-ml-federated-learning/blob/thomas/pneumonia-xray-example/docs/real-world-examples/pneumonia.md) doc.

> Note: It is not expected of you to do the whole testing described above as part of the code review. This is more for the bug bash. For the code review, it is totally OK if you just read the code and not actually run it. But if you do want to run it, don't let this note stop you ;)

## What to Check
Verify that the following are valid:
* the data are split properly into the 3 silos;
* the FL job runs successfully.

## Follow-up work
- [ ] use a shared component for model aggregation
- [ ] only keep one environment for all examples
- [ ] add this experiment to the CI/CD?
- [ ] revisit the story around data provisioning
